### ISSUE TITLE
`amplify push` doesn't read from `backend-config`, but does read from `amplify-meta`

### ISSUE THREAD

**Issue by nagey** [aws-amplify/amplify-cli#3426]
**Describe the bug**
When running `amplify push` I expect it to consult the `backend-config.json` file, especially the `dependsOn` array for various services.

Apparently though, running `amplify env checkout …` is the magic sauce to convert the items in backend-config into amplify-meta. 

**Amplify CLI Version**
4.13.1

**To Reproduce**
1. Make a change to your backend-config.json. 
1. Push to the cloud. 
1. Stare at your cloudformation stacks in the console, wonder why change isn't being reflected
1. Deliberately break backend-config.json
1. push to the cloud. not broken.
1. google a bunch of things
1. get frustrated
1. have dinner
1. come back, and repeat steps 1-7
1. On a whim, look at `amplify-meta.json` and see that while it's out of sync with your `backend-config.json` it's strangely in sync with what seems to be happening in the CF Template.
1. Finally have someone point out to you the 4th item #1 under https://aws-amplify.github.io/docs/cli-toolchain/quickstart#custom-cloudformation-stacks and see the bit about having to run `amplify env checkout …` to get amplify to read the changes on your local system

**Expected behavior**
The build processes associated with amplify push should read `backend-config` in a similar fashion to how it reads all of the various cloudformation templates, schemas, functions, and everything else that're kept in source control.

**Additional context**
I would really like to get these three hours back, but they're gone forever, sitting contently in a nice crowded room with all of the other (numerous) hours that I've lost to poorly documented, or buggy, pieces of the amplify codebase.

**Comment by nagey** [comment]
Preemptively sorry for my semi sarcastic tone in my steps to repro, but this was an extremely frustrating thing to find, and honestly makes no sense to me as to why it is done this way. 

If there’s a compelling reason that it is the way it is, could there at least be a comment put into the backend config that it only gets read when you do an env pull?

**Comment by ammarkarachi** [comment]
Hi @nagey we generally don't recommend modifying those files manually. Can you give us more information on what you were trying to do?

**Comment by nagey** [comment]
"generally don't recommend"?
![image](https://user-images.githubusercontent.com/13761/74564322-cd58c380-4f3c-11ea-82e6-82d5af970cf1.png)

Unless you mean modifying amplify-meta, which, yeah, I get. I was trying to pipe the endpoint URL of one of the rest services into a function, rather than just the endpoints "friendly" name.

**Comment by kaustavghosh06** [comment]
@nagey Yes, the amplify-meta.json are runtime files which should not be modified. But yes, you're right the `backend-config.json` has to be modified for adding custom resources/categories. And yes, as mentioned in step #4 of our documentation, we've mentioned to run `amplify env checkout <env-name>` after making any changes to your backend-config.json. But at the same time, I agree this could be a part of your `amplify push` process to make the experience better.

**Comment by nagey** [comment]
@kaustavghosh06 I've made sure to add the checkout step to our workflow when changing resources, but would love to hear it getting added to the amplify push

**Comment by asmajlovicmars** [comment]
wow. thanks for this... spent hours trying to figure out why suddenly adding sns to backend-config doesn't work... lol

goes into readme: **run amplify env checkout <env-name> after making any changes to your backend-config.json**

**Comment by irfan** [comment]
I have had same issue today, even you don't have to go to aws console, you can see in your terminal output that there is nothing to push.

I executed `amplify env checkout irfansenv` and then tried `amplify push` again, then amplify finally asked me if I want to continue.

**Comment by AndreasEK** [comment]
Please, please, please add it to the ' amplify push' flow. I do have a reminder to run `amplify env pull` before an `amplify push`- but I still keep forgetting about it and need more time than it should to figure out why things don't work. THANKS <3

**Comment by ldaudet** [comment]
Hi there!

This is still very much an issue. Lost hours trying to figure it out. It makes no sense that we have to run amplify env checkout every time we push. At the beginning of the amplify push, it absolutely should do that comparison and update amplify meta based on the content of the backend config.json.

Thanks

**Comment by a9udn9u** [comment]
I have a custom stack to take care of assigning permissions to my Cognito user groups and Lambda functions. This issue bit me 6 months ago, today I added a new Lambda function and it bites me again. I've lost at least a day because of this...

`backend-config.json` is not a file developers update everyday, WHO ON EARTH WOULD REMEMBER TO RUN `amplify env checkout` AFTER SIX MONTHS??!!

`amplify push` should update `amplify-meta.json` automatically! Does it make sense for anyone to update `backend-config.json` but not want the change to take effect the next time they push?

**Comment by pszabop** [comment]
I just burned 6 hours on similar problem, it was a badly performd merge (from a conflict) on backend-config.json that got me.

amplify push should rebuild *everything* and not leave some changes unbuilt.

**Comment by dral3x** [comment]
> [...] But at the same time, I agree this could be a part of your amplify push process to make the experience better.
@kaustavghosh06 any update on this? 2 years after your messages and `amplify-meta.json` is still not updated after `amplify push`

**Comment by Jeevantk** [comment]
I just burned 8 hours on this issue. Can we please prioritize to update amplify-meta.json on `amplify push`

**Comment by JosueBCe** [comment]
Is it not about this? https://github.com/aws-amplify/amplify-cli/issues/9941#issuecomment-1063231806

**Comment by JosueBCe** [comment]
Here there's a solution https://github.com/aws-amplify/amplify-cli/issues/8608#issuecomment-1086890943

**Comment by JosueBCe** [comment]
> I just burned 8 hours on this issue. Can we please prioritize updating
Worked for me @Jeevantk 
https://github.com/aws-amplify/amplify-cli/issues/8608#issuecomment-1086890943 amplify-meta.json on `amplify push`
